File: 2024/day12/main.c
#include <stdio.h>
#include <stdlib.h>
#include <string.h>

#define BUFFER_SIZE 512

typedef struct {
  int x[2], y[2];
} Fence;

typedef struct {
  int nx, ny;
  char **grid;
} Plots;

Plots *createPlots(int n_row, int n_col, char *pixels) {
  Plots *plots = (Plots *)malloc(sizeof(Plots));
  plots->ny = n_row;
  plots->nx = n_col;
  plots->grid = (char **)malloc(sizeof(char *) * n_row);
  plots->grid[0] = pixels;
  for (int i = 1; i < n_row; ++i) {
    plots->grid[i] = plots->grid[i - 1] + n_col;
  }
  return plots;
}

void freePlots(Plots *plots) {
  free(plots->grid[0]);
  free(plots->grid);
  free(plots);
}

void setFencePoints(Fence *fence, int x, int y, char side) {
  switch (side) {
  case 'N':
    fence->x[0] = x + 1;
    fence->y[0] = y;
    fence->x[1] = x;
    fence->y[1] = y;
    break;
  case 'W':
    fence->x[0] = x;
    fence->y[0] = y;
    fence->x[1] = x;
    fence->y[1] = y + 1;
    break;
  case 'S':
    fence->x[0] = x;
    fence->y[0] = y + 1;
    fence->x[1] = x + 1;
    fence->y[1] = y + 1;
    break;
  case 'E':
    fence->x[0] = x + 1;
    fence->y[0] = y + 1;
    fence->x[1] = x + 1;
    fence->y[1] = y;
    break;
  default:
    printf("Invalid side\n");
    exit(1);
  }
}

int connected(const Fence *f0, const Fence *f1) {
  return f0->x[1] == f1->x[0] && f0->y[1] == f1->y[0];
}

int isCorner(const Fence *f0, const Fence *f1) {
  int u0 = f0->x[1] - f0->x[0];
  int v0 = f0->y[1] - f0->y[0];
  int u1 = f1->x[1] - f1->x[0];
  int v1 = f1->y[1] - f1->y[0];
  return u0 != u1 || v0 != v1;
}

void computePrice(const Plots *plots, const char region, int x, int y,
                  int *visited, int *area, int *perimeter, Fence *fences) {
  visited[y * plots->nx + x] = 1;
  *area += 1;
  int shiftx[4] = {0, 1, 0, -1};
  int shifty[4] = {1, 0, -1, 0};
  char side[4] = {'S', 'E', 'N', 'W'};
  int x_, y_;
  for (int k = 0; k < 4; ++k) {
    x_ = x + shiftx[k];
    y_ = y + shifty[k];
    if (x_ < 0 || x_ >= plots->nx || y_ < 0 || y_ >= plots->ny ||
        plots->grid[y_][x_] != region) {
      setFencePoints(fences + *perimeter, x, y, side[k]);
      *perimeter += 1;
      continue;
    }
    if (!visited[y_ * plots->nx + x_])
      computePrice(plots, region, x_, y_, visited, area, perimeter, fences);
  }
}

int countCorners(Fence *fences, int perimeter) {
  int corners = 0;
  int icurr = 0;
  Fence *first = &fences[icurr];
  Fence tmp, *curr, *next;
  curr = first;
  while (icurr < perimeter) {
    for (int i = icurr + 1; i < perimeter; ++i) {
      next = &fences[i];
      if (connected(curr, next)) {
        if (isCorner(curr, next))
          corners++;
        icurr += 1;
        tmp = fences[icurr];
        fences[icurr] = fences[i];
        fences[i] = tmp;
        curr = &fences[icurr];
        break;
      }
    }
    if (connected(curr, first)) {
      if (isCorner(curr, first))
        corners++;
      icurr += 1;
      first = &fences[icurr];
      curr = first;
    }
  }
  return corners;
}

int main(int argc, char *argv[]) {
  if (argc != 2) {
    printf("Missing input file\n");
    return 1;
  }

  FILE *file = fopen(argv[1], "r");
  if (file == NULL) {
    printf("Could not open file\n");
    return 1;
  }
  char *gardens = (char *)malloc(sizeof(char) * BUFFER_SIZE * BUFFER_SIZE);
  char buffer[BUFFER_SIZE];
  int n_col, n_row = 0;
  while (fgets(buffer, BUFFER_SIZE, file)) {
    n_col = strcspn(buffer, "\n");
    for (int i = 0; i < n_col; ++i) {
      gardens[n_row * n_col + i] = buffer[i];
    }
    n_row++;
  }
  fclose(file);

  gardens = (char *)realloc(gardens, sizeof(char) * n_row * n_col);
  int *visited = (int *)calloc(sizeof(int), n_row * n_col);
  Fence *fences = (Fence *)malloc(sizeof(Fence) * n_row * n_col);
  Plots *plots = createPlots(n_row, n_col, gardens);

  int area, perimeter, corners;
  int part1 = 0, part2 = 0;
  for (int y = 0; y < plots->ny; ++y) {
    for (int x = 0; x < plots->nx; ++x) {
      if (visited[y * plots->nx + x]) {
        continue;
      }
      perimeter = area = 0;
      computePrice(plots, plots->grid[y][x], x, y, visited, &area, &perimeter,
                   fences);
      part1 += area * perimeter;
      corners = countCorners(fences, perimeter);
      part2 += area * corners;
    }
  }

  printf("Part 1: %d\n", part1);
  printf("Part 2: %d\n", part2);

  freePlots(plots);
  free(visited);
  free(fences);
  return 0;
}

File: 2024/day12/src/main.rs
use anyhow::Result;
use std::env::args;
use std::fs::File;
use std::io::{BufRead, BufReader};

enum Direction {
    North,
    East,
    South,
    West,
}

struct Fence {
    a: (usize, usize),
    b: (usize, usize),
}

impl Fence {
    pub fn as_vector(&self) -> (isize, isize) {
        let (x1, y1) = self.a;
        let (x2, y2) = self.b;
        (x2 as isize - x1 as isize, y2 as isize - y1 as isize)
    }
}

fn is_corner(f0: &Fence, f1: &Fence) -> bool {
    let (u0, v0) = f0.as_vector();
    let (u1, v1) = f1.as_vector();
    u0 != u1 || v0 != v1
}

fn get_neighbour(
    plots: &[Vec<char>],
    x: usize,
    y: usize,
    side: &Direction,
) -> Option<(usize, usize)> {
    let (x_, y_) = match side {
        Direction::North => (x, y.checked_sub(1)?),
        Direction::East => (x + 1, y),
        Direction::South => (x, y + 1),
        Direction::West => (x.checked_sub(1)?, y),
    };
    if x_ >= plots[0].len() || y_ >= plots.len() {
        None
    } else if plots[y_][x_] != plots[y][x] {
        None
    } else {
        Some((x_, y_))
    }
}

fn fence_from_side(x: usize, y: usize, side: &Direction) -> Fence {
    match side {
        Direction::North => Fence {
            a: (x + 1, y),
            b: (x, y),
        },
        Direction::West => Fence {
            a: (x, y),
            b: (x, y + 1),
        },
        Direction::South => Fence {
            a: (x, y + 1),
            b: (x + 1, y + 1),
        },
        Direction::East => Fence {
            a: (x + 1, y + 1),
            b: (x + 1, y),
        },
    }
}

fn compute_price(
    plots: &[Vec<char>],
    visited: &mut [Vec<bool>],
    x: usize,
    y: usize,
    area: &mut usize,
    perimeter: &mut usize,
    fences: &mut Vec<Fence>,
) {
    visited[y][x] = true;
    *area += 1;
    let sides = [
        Direction::North,
        Direction::East,
        Direction::South,
        Direction::West,
    ];
    for side in sides.iter() {
        let (x_, y_) = match get_neighbour(plots, x, y, side) {
            Some((x_, y_)) => (x_, y_),
            None => {
                *perimeter += 1;
                fences.push(fence_from_side(x, y, side));
                continue;
            }
        };
        if !visited[y_][x_] {
            compute_price(plots, visited, x_, y_, area, perimeter, fences);
        }
    }
}

fn count_corners(fences: &mut [Fence]) -> usize {
    let mut corners = 0;
    let mut icurr = 0;
    let mut first = 0;
    while icurr < fences.len() {
        for i in icurr + 1..fences.len() {
            if fences[icurr].b == fences[i].a {
                if is_corner(&fences[icurr], &fences[i]) {
                    corners += 1;
                }
                fences.swap(icurr + 1, i);
                icurr += 1;
                break;
            }
        }
        if fences[first].a == fences[icurr].b {
            if is_corner(&fences[first], &fences[icurr]) {
                corners += 1;
            }
            icurr += 1;
            first = icurr;
        }
    }
    corners
}

fn main() -> Result<()> {
    let filename = args().nth(1).expect("Missing input file");
    let file = File::open(filename)?;
    let reader = BufReader::new(file);
    let mut plots: Vec<Vec<char>> = Vec::new();
    for line in reader.lines() {
        let line = line?;
        plots.push(line.chars().collect());
    }

    let nx = plots[0].len();
    let ny = plots.len();
    // println!("nx: {}, ny: {}", nx, ny);
    let mut visited = vec![vec![false; nx]; ny];
    let mut part1 = 0;
    let mut part2 = 0;
    for y in 0..ny {
        for x in 0..nx {
            if visited[y][x] {
                continue;
            }
            let mut area = 0;
            let mut perimeter = 0;
            let mut fences = Vec::new();
            compute_price(
                &plots,
                &mut visited,
                x,
                y,
                &mut area,
                &mut perimeter,
                &mut fences,
            );
            part1 += area * perimeter;
            part2 += area * count_corners(&mut fences);
        }
    }

    println!("Part 1: {}", part1);
    println!("Part 2: {}", part2);

    Ok(())
}
